### PR TITLE
Add Expense Agent E2E test setup API

### PR DESCRIPTION
**Pull request by t-prda**
## Summary

- add an idempotent, ServiceEnabled `Configure` action for Expense Agent service E2E setup
- create required wizard defaults and master data before creating or reusing an enabled Expense Agent
- keep `Initialize` cleanup-only and remove transactional data plus stale Expense Vendors between E2E tests
- support tenant-routed test-handler calls without registering unsupported web services

## Scope

This PR provides the BC test-handler surface required by the Expense Agent service E2E pipeline. New AL API test scenarios are intentionally deferred to [AB#636056](https://dynamicssmb2.visualstudio.com/1fcb79e7-ab07-432a-a3c6-6cf5a88ba4a5/_workitems/edit/636056), which owns API-test conversion, permissions, capability/privacy/API-auth setup, and runner isolation.

## Companion change

- Expense Agent pipeline: https://microsoft.ghe.com/bic/BC-ExpenseAgent/pull/2404
- NAV integration branch: `private/prdas/expense-agent-e2e-pipeline`
- Deliverable: [AB#636054](https://dynamicssmb2.visualstudio.com/1fcb79e7-ab07-432a-a3c6-6cf5a88ba4a5/_workitems/edit/636054)

## Validation

- branch rebased on current BCApps `main`
- `git diff --check` passes
- hosted aligned BC30 run 2704842 succeeded against BC artifact `30.0.54034.0/w1`
- 146 ERP integration tests collected: **132 passed, 14 skipped, 0 failed**
- repeated setup preserves tenant routing, reuses existing setup, and rejects disabled underlying Agent state

### CODE REVIEW

File: src/Apps/W1/ExpenseAgent/test/src/Helper/ExpenseTestHandlerAPI.Codeunit.al [hunk 1]
```diff
@@ -10,22 +10,112 @@ using Microsoft.Projects.Project.Job;
...
+        exit(AgentUserSecurityId);
+    end;
+
+    local procedure EnableExpenseAgentSetup(
+        var ExpenseAgentSetup: Record "Expense Agent Setup";
+        AgentUserSecurityId: Guid)
+    begin
+        if not ExpenseAgentSetup.Get() then
+            ExpenseAgentSetup.InitRecord();
+        ExpenseAgentSetup."User Security ID" := AgentUserSecurityId;
```
[the remark this answers is not in the document]
Comment (t-prda) [reply src/Apps/W1/ExpenseAgent/test/src/Helper/ExpenseTestHandlerAPI.Codeunit.al]: System.Collections.Hashtable.b

File: src/Apps/W1/ExpenseAgent/test/src/Helper/ExpenseTestHandlerAPI.Codeunit.al [hunk 2]
```diff
@@ -10,22 +10,112 @@ using Microsoft.Projects.Project.Job;
...
+
+        CreateExpenseAgentSetup.Run();
+        CreateExpenseGLAccount.Run();
+        EnsureExpenseAgent();
+        ExpenseAgentSetup.Get();
+        ExpenseAgentSetup.CreateDefaultSettings();
         exit('Initialize completed');
     end;
 
+    local procedure ExpenseAgentSetupIsComplete(): Boolean
```
[the remark this answers is not in the document]
Comment (t-prda) [reply src/Apps/W1/ExpenseAgent/test/src/Helper/ExpenseTestHandlerAPI.Codeunit.al:53]: System.Collections.Hashtable.b

File: src/Apps/W1/ExpenseAgent/test/src/API/ExpenseAgentSetupAPITest.Codeunit.al [hunk 3]
```diff
@@ -70,4 +202,72 @@ codeunit 148333 "Expense Agent Setup API Test"
...
+        Assert.IsTrue(NoSeries.Get(ExpenseAgentSetup."Expense Reports Nos."), 'The expense report number series must exist.');
+        Assert.IsTrue(NoSeries.Get(ExpenseAgentSetup."Posted Expense Reports Nos."), 'The posted expense report number series must exist.');
+    end;
+
+    local procedure VerifyDefaultPostingGroupAccountsExist()
+    var
+        ExpensePostingGroup: Record "Expense Posting Group";
+        GLAccount: Record "G/L Account";
+    begin
+        Assert.IsTrue(ExpensePostingGroup.FindSet(), 'At least one default expense posting group must exist.');
```
[the remark this answers is not in the document]
Comment (t-prda) [reply src/Apps/W1/ExpenseAgent/test/src/API/ExpenseAgentSetupAPITest.Codeunit.al]: System.Collections.Hashtable.b

File: src/Apps/W1/ExpenseAgent/test/src/Helper/ExpenseTestHandlerAPI.Codeunit.al [hunk 4]
```diff
@@ -10,22 +10,112 @@ using Microsoft.Projects.Project.Job;
...
+    end;
+
     /// <summary>
     /// Resets expense transactional data so tests can start from a clean slate.
     /// Example: curl.exe -X POST "http://localhost:7047/Navision_NAV/ODataV4/ExpenseTestHandler_Initialize?company=4d77b30a-3dc5-f011-9c12-7ced8d9e45e7" -u username:password -H "Content-Type: application/json" -H "Accept: application/json" -d "{}"
     /// </summary>
     [ServiceEnabled]
     procedure Initialize(): Text[30]
     var
+        ExpenseAgentSetup: Record "Expense Agent Setup";
```
[the remark this answers is not in the document]
Comment (t-prda) [reply src/Apps/W1/ExpenseAgent/test/src/Helper/ExpenseTestHandlerAPI.Codeunit.al:37]: System.Collections.Hashtable.b

File: src/Apps/W1/ExpenseAgent/test/src/API/ExpenseAgentSetupAPITest.Codeunit.al [hunk 5]
```diff
@@ -58,6 +62,134 @@ codeunit 148333 "Expense Agent Setup API Test"
...
+    var
+        ExpenseAgentSetup: Record "Expense Agent Setup";
+        Agent: Record Agent;
+        ExpenseTestHandlerAPI: Codeunit "Expense Test Handler API";
+        FirstAgentUserSecurityId: Guid;
+        SecondAgentUserSecurityId: Guid;
+    begin
+        // [FEATURE] [AI test 1.0]
+        // [SCENARIO] Test handler initialization creates and reuses an enabled Expense Agent
+        Initialize();
```
[the remark this answers is not in the document]
Comment (t-prda) [reply src/Apps/W1/ExpenseAgent/test/src/API/ExpenseAgentSetupAPITest.Codeunit.al]: System.Collections.Hashtable.b

File: src/Apps/W1/ExpenseAgent/test/src/Helper/ExpenseTestHandlerAPI.Codeunit.al [hunk 6]
```diff
@@ -10,22 +10,112 @@ using Microsoft.Projects.Project.Job;
...
+
+        EnableExpenseAgentSetup(ExpenseAgentSetup, AgentUserSecurityId);
+        exit(AgentUserSecurityId);
+    end;
+
+    local procedure EnableExpenseAgentSetup(
+        var ExpenseAgentSetup: Record "Expense Agent Setup";
+        AgentUserSecurityId: Guid)
+    begin
+        if not ExpenseAgentSetup.Get() then
```
[the remark this answers is not in the document]
Comment (t-prda) [reply src/Apps/W1/ExpenseAgent/test/src/Helper/ExpenseTestHandlerAPI.Codeunit.al]: System.Collections.Hashtable.b

File: src/Apps/W1/ExpenseAgent/test/src/API/ExpenseAgentSetupAPITest.Codeunit.al [hunk 7]
```diff
@@ -58,6 +62,137 @@ codeunit 148333 "Expense Agent Setup API Test"
...
+    var
+        ExpenseVendor: Record "Expense Vendor";
+        ExpenseTestHandlerAPI: Codeunit "Expense Test Handler API";
+    begin
+        // [FEATURE] [AI test 1.0]
+        // [SCENARIO] Test handler initialization removes stale Expense Vendors
+        Initialize();
+
+        // [GIVEN] An Expense Vendor exists from a previous test
+        ExpenseVendor.Init();
```
[the remark this answers is not in the document]
Comment (t-prda) [reply src/Apps/W1/ExpenseAgent/test/src/API/ExpenseAgentSetupAPITest.Codeunit.al]: Fixed in ad2e3a1f4. `Library - Expense` now exposes an internal `CreateExpenseVendor` helper that inserts through the table trigger and configured number series; the cleanup test uses that valid fixture and its generated key.

File: src/Apps/W1/ExpenseAgent/test/src/Helper/ExpenseTestHandlerAPI.Codeunit.al [hunk 8]
```diff
@@ -10,22 +10,112 @@ using Microsoft.Projects.Project.Job;
...
+        ExpenseAgentSetup: Record "Expense Agent Setup";
+        CreateExpenseAgentSetup: Codeunit "Create Expense Agent Setup";
+        CreateExpenseGLAccount: Codeunit "Create Expense GL Account";
         LibraryExpense: Codeunit "Library - Expense";
     begin
         LibraryExpense.CleanTransactionalData();
+        if ExpenseAgentSetupIsComplete() then
+            exit('Initialize completed');
+
+        CreateExpenseAgentSetup.Run();
```
[the remark this answers is not in the document]
Comment (t-prda) [reply src/Apps/W1/ExpenseAgent/test/src/Helper/ExpenseTestHandlerAPI.Codeunit.al:45]: System.Collections.Hashtable.b

File: src/Apps/W1/ExpenseAgent/test/src/Helper/ExpenseTestHandlerAPI.Codeunit.al [hunk 9]
```diff
@@ -10,22 +10,112 @@ using Microsoft.Projects.Project.Job;
...
+    end;
+
     /// <summary>
     /// Resets expense transactional data so tests can start from a clean slate.
     /// Example: curl.exe -X POST "http://localhost:7047/Navision_NAV/ODataV4/ExpenseTestHandler_Initialize?company=4d77b30a-3dc5-f011-9c12-7ced8d9e45e7" -u username:password -H "Content-Type: application/json" -H "Accept: application/json" -d "{}"
     /// </summary>
     [ServiceEnabled]
     procedure Initialize(): Text[30]
     var
+        ExpenseAgentSetup: Record "Expense Agent Setup";
```
[the remark this answers is not in the document]
Comment (t-prda) [reply src/Apps/W1/ExpenseAgent/test/src/Helper/ExpenseTestHandlerAPI.Codeunit.al:37]: System.Collections.Hashtable.b

File: src/Apps/W1/ExpenseAgent/test/src/API/ExpenseAgentSetupAPITest.Codeunit.al [hunk 10]
```diff
@@ -58,6 +62,139 @@ codeunit 148333 "Expense Agent Setup API Test"
...
+        Assert.IsTrue(ExpensePaymentMethod.Get('CARD'), 'The credit-card payment method must exist.');
+        Assert.IsTrue(ExpenseCategory.Get('MEALS'), 'The meals category must exist.');
+        Assert.IsTrue(ExpenseCategory.Get('HOTELS'), 'The hotels category must exist.');
+        Assert.IsTrue(ExpenseCategory.Get('ENTERTAIN'), 'The entertainment category must exist.');
+        Assert.IsTrue(ExpenseCategory.Get('PER-DIEM'), 'The per-diem category must exist.');
+        VerifyDefaultPostingGroupAccountsExist();
+    end;
+
+    [Test]
+    procedure TestHandlerInitializeIsIdempotent()
```
[the remark this answers is not in the document]
Comment (t-prda) [reply src/Apps/W1/ExpenseAgent/test/src/API/ExpenseAgentSetupAPITest.Codeunit.al]: System.Collections.Hashtable.b

File: src/Apps/W1/ExpenseAgent/test/src/API/ExpenseAgentSetupAPITest.Codeunit.al [hunk 11]
```diff
@@ -70,4 +207,80 @@ codeunit 148333 "Expense Agent Setup API Test"
...
+        ExpensePostingGroup: Record "Expense Posting Group";
+        ExpenseSubcategory: Record "Expense Subcategory";
+    begin
+        LibraryExpense.CleanTransactionalData();
+        ExpenseSubcategory.DeleteAll();
+        ExpenseCategory.DeleteAll();
+        ExpensePaymentMethod.DeleteAll();
+        ExpensePostingGroup.DeleteAll();
+
+        ExpenseAgentSetup.Get();
```
[the remark this answers is not in the document]
Comment (t-prda) [reply src/Apps/W1/ExpenseAgent/test/src/API/ExpenseAgentSetupAPITest.Codeunit.al]: System.Collections.Hashtable.b

File: src/Apps/W1/ExpenseAgent/test/src/API/ExpenseAgentSetupAPITest.Codeunit.al [hunk 12]
```diff
@@ -58,6 +62,139 @@ codeunit 148333 "Expense Agent Setup API Test"
...
+        ExpenseTestHandlerAPI.Initialize();
+        ExpenseAgentSetup.Get();
+        SecondAgentUserSecurityId := ExpenseAgentSetup."User Security ID";
+
+        // [THEN] The same enabled Expense Agent is reused
+        Assert.AreEqual(
+            FirstAgentUserSecurityId,
+            SecondAgentUserSecurityId,
+            'Repeated E2E initialization must reuse the same Expense Agent.');
+        Assert.IsTrue(Agent.Get(FirstAgentUserSecurityId), 'The Expense Agent must exist.');
```
[the remark this answers is not in the document]
Comment (t-prda) [reply src/Apps/W1/ExpenseAgent/test/src/API/ExpenseAgentSetupAPITest.Codeunit.al]: System.Collections.Hashtable.b

File: src/Apps/W1/ExpenseAgent/test/src/Helper/ExpenseTestHandlerAPI.Codeunit.al (hunk 9 above)
[the remark this answers is not in the document]
Comment (t-prda) [reply src/Apps/W1/ExpenseAgent/test/src/Helper/ExpenseTestHandlerAPI.Codeunit.al:37]: System.Collections.Hashtable.b

File: src/Apps/W1/ExpenseAgent/test/src/Helper/ExpenseTestHandlerAPI.Codeunit.al [hunk 13]
```diff
@@ -26,6 +81,51 @@ codeunit 148307 "Expense Test Handler API"
         exit('Initialize completed');
     end;
 
+    /// <summary>
+    /// Explicitly provisions and enables Expense Agent setup for E2E tests.
+    /// </summary>
+    [ServiceEnabled]
+    procedure Configure(): Text[30]
```
[the remark this answers is not in the document]
Comment (t-prda) [reply src/Apps/W1/ExpenseAgent/test/src/Helper/ExpenseTestHandlerAPI.Codeunit.al:35]: System.Collections.Hashtable.b

File: src/Apps/W1/ExpenseAgent/test/src/API/ExpenseAgentSetupAPITest.Codeunit.al [hunk 14]
```diff
@@ -58,6 +62,152 @@ codeunit 148333 "Expense Agent Setup API Test"
...
+        // [SCENARIO] Repeated test handler initialization reuses completed setup data
+        Initialize();
+
+        // [GIVEN] Required Expense Agent setup data is absent
+        ClearRequiredSetupData();
+
+        // [GIVEN] The E2E test handler initializes Expense Agent master data
+        ExpenseTestHandlerAPI.Configure();
+        ExpenseAgentSetup.Get();
+        CategoryCount := ExpenseCategory.Count();
```
[the remark this answers is not in the document]
Comment (t-prda) [reply src/Apps/W1/ExpenseAgent/test/src/API/ExpenseAgentSetupAPITest.Codeunit.al]: System.Collections.Hashtable.b

**Comment by alexei-dobriansky** [comment]
## Good Sense Reviewer - Round 1

# **Recommendation: Request Changes**

<details>
<summary><strong>What this PR does</strong></summary>

This change splits the test handler into a cleanup action and a setup action. The setup action creates default Expense Agent data, creates or reuses an Agent record, and adds tests for first setup, repeated setup, default data, and stale vendor cleanup.

Most of the setup order is correct: default data is created before validating `"Enable Agent"`, and the tests cover the common fresh-company path. One state update path still leaves an existing disabled Agent inactive while the setup action reports success.
</details>

<details>
<summary><strong>Problem-solution fit</strong></summary>

**Fit:** Partial

The requested behavior is clear and the new API mostly matches it. It does not fully handle a company where the Expense Agent record already exists but is disabled, so the setup can still be incomplete after the call returns success.
</details>

<details>
<summary><strong>Suggestions</strong></summary>

**S1** (🔴 High): *Existing disabled Agent stays disabled*
The code assigns `TempAgentSetupBuffer.State` directly before `SaveChanges`. The buffer only marks state changes when the field is validated, so `SaveChanges` can skip `Agent.Activate()` for an existing disabled Agent and still set `"Enable Agent"` to true. Validate the `State` field and also make the complete check reject a disabled Agent before returning success.
</details>

<details>
<summary><strong>Risk assessment and necessity</strong></summary>

**Risk:** The changed code is test setup infrastructure, not posting logic, so the financial regression surface is low. The main risk is false E2E stability: a company can look configured while the underlying Agent is inactive, which can make later agent flows fail for the wrong reason.

**Necessity:** The change is needed to make Expense Agent E2E setup repeatable and service-driven. The scope is reasonable, but the setup action must make both setup records and the underlying Agent state consistent.
</details>

---
> [AI-PR-REVIEW] version=1 promptVersion=4 system=github pr=10846 round=1 by=alexei-dobriansky at=2026-09-01T01:18:40.1472128Z lastSha=790fd6602bb5c88fac8fb2dbb702f8ee7e271dd6 reviewKey=cbcee2444c10619a44ec73088c3abe62646f6b7692f57ca45298cecc5bc1cbca suggestions=S1@f54b7baa

File: src/Apps/W1/ExpenseAgent/test/src/API/ExpenseAgentSetupAPITest.Codeunit.al [hunk 15]
```diff
@@ -70,4 +266,86 @@ codeunit 148333 "Expense Agent Setup API Test"
...
+            Assert.IsTrue(
+                GLAccount.Get(ExpensePostingGroup."Debit Rounding Account"),
+                'The debit rounding account must exist.');
+            Assert.IsTrue(
+                GLAccount.Get(ExpensePostingGroup."Credit Rounding Account"),
+                'The credit rounding account must exist.');
+        until ExpensePostingGroup.Next() = 0;
+    end;
+
+    [ConfirmHandler]
```
[the remark this answers is not in the document]
Comment (t-prda) [reply src/Apps/W1/ExpenseAgent/test/src/API/ExpenseAgentSetupAPITest.Codeunit.al]: Fixed in a30cc87fe. Each configuration test queues the expected `false` reply, the handler records the actual question, and the test verifies the exact prompt plus `LibraryVariableStorage.AssertEmpty()`.

**Comment by alexei-dobriansky** [comment]
## Good Sense Reviewer - Round 2

# **Recommendation: Request Changes**

<details>
<summary><strong>What this PR does</strong></summary>

The new commits add confirmation handling for the setup tests that run default Expense Agent setup. The prompt is now handled and verified, so those tests can exercise the setup flow without an unhandled confirmation.

That change is correct for the new prompt, but it does not change the setup API path that was reviewed in round 1. The API still assigns the Agent setup buffer state directly before saving, and the completion check still only checks that the Agent record exists.
</details>

<details>
<summary><strong>Status of previous suggestions</strong></summary>

| ID | Title | Status | Author response |
|----|-------|--------|-----------------|
| S1 | Existing disabled Agent stays disabled | Not addressed | No response. The current code still assigns TempAgentSetupBuffer.State directly and still does not require the existing Agent to be enabled before returning success. |
</details>

<details>
<summary><strong>New observations (commits since round 1)</strong></summary>

*None - the new commits only add confirmation handling in tests and do not introduce another issue.*
</details>

<details>
<summary><strong>Risk assessment and necessity</strong></summary>

**Risk:** The changed surface is E2E setup infrastructure, not posting logic, so the direct financial risk is low. The remaining risk is false setup success: later agent flows can run in a company where setup says the agent is enabled but the underlying Agent is still inactive.

**Necessity:** The setup API is needed for repeatable service-driven E2E tests. The added confirmation handling is also needed, but the setup API still must make the Expense Agent setup record and the underlying Agent state agree.
</details>

---
> [AI-PR-REVIEW] version=1 promptVersion=4 system=github pr=10846 round=2 by=alexei-dobriansky at=2026-09-01T07:18:05.4760372Z lastSha=a30cc87fe83c128a779837db364e2428313605d9 reviewKey=323acf5817c3301562c7bc5b0ca1bc570777023623d6eb7e091bad7c2c60090c suggestions=S1@f54b7baa:notaddressed parentRound=1

**Review by Alexander-Ya — DISMISSED** [review]

**Comment by t-prda** [comment]
Addressed the regular review comments in `636c173e1`:

- Good Sense S1: Agent buffer state now uses `Validate(State, Enabled)`, completeness rejects a disabled Agent, and a dedicated test proves an existing disabled Agent is re-enabled.
- Confirmation handling: tests now verify the exact prompt and that it fires exactly once.
- `CreateExpenseVendor`: the internal library helper now has XML documentation.

The endpoint-specific not-found diagnostics and lower-camel parameter comments refer to procedures already present on `main`, not changes introduced by this PR. Renaming ServiceEnabled parameters would also change the OData action payload contract, so they are intentionally unchanged here.

File: src/Apps/W1/ExpenseAgent/test/src/Helper/ExpenseTestHandlerAPI.Codeunit.al [hunk 16]
```diff
@@ -26,6 +81,63 @@ codeunit 148307 "Expense Test Handler API"
...
+    var
+        ExpenseAgentSetup: Record "Expense Agent Setup";
+        CreateExpenseAgentSetup: Codeunit "Create Expense Agent Setup";
+        CreateExpenseGLAccount: Codeunit "Create Expense GL Account";
+    begin
+        if ExpenseAgentSetupIsComplete() then
+            exit('Configure completed');
+
+        CreateExpenseAgentSetup.Run();
```
[the remark this answers is not in the document]
Comment (t-prda) [reply src/Apps/W1/ExpenseAgent/test/src/Helper/ExpenseTestHandlerAPI.Codeunit.al:45]: System.Collections.Hashtable.b

File: src/Apps/W1/ExpenseAgent/test/src/API/ExpenseAgentSetupAPITest.Codeunit.al [hunk 17]
```diff
@@ -41,6 +49,56 @@ codeunit 148333 "Expense Agent Setup API Test"
             ExpenseAgentSetup.Init();
             ExpenseAgentSetup.Insert();
         end;
+
+        [Test]
```
[the remark this answers is not in the document]
Comment (t-prda) [reply src/Apps/W1/ExpenseAgent/test/src/API/ExpenseAgentSetupAPITest.Codeunit.al]: System.Collections.Hashtable.b

File: src/Apps/W1/ExpenseAgent/test/src/API/ExpenseAgentSetupAPITest.Codeunit.al [hunk 18]
```diff
@@ -70,4 +326,110 @@ codeunit 148333 "Expense Agent Setup API Test"
...
+    begin
+        ExpenseAgentSetup.Get();
+        Assert.IsTrue(NoSeries.Get(ExpenseAgentSetup."Expense Nos."), 'The expense number series must exist.');
+        Assert.IsTrue(NoSeries.Get(ExpenseAgentSetup."Expense User Nos."), 'The expense user number series must exist.');
+        Assert.IsTrue(NoSeries.Get(ExpenseAgentSetup."Expense Vendor Nos."), 'The expense vendor number series must exist.');
+        Assert.IsTrue(NoSeries.Get(ExpenseAgentSetup."Expense Reports Nos."), 'The expense report number series must exist.');
+        Assert.IsTrue(NoSeries.Get(ExpenseAgentSetup."Posted Expense Reports Nos."), 'The posted expense report number series must exist.');
+    end;
+
+    local procedure EnqueueCopyEmployeesConfirmation()
```
[the remark this answers is not in the document]
Comment (t-prda) [reply src/Apps/W1/ExpenseAgent/test/src/API/ExpenseAgentSetupAPITest.Codeunit.al]: System.Collections.Hashtable.b

File: src/Apps/W1/ExpenseAgent/test/src/API/ExpenseAgentSetupAPITest.Codeunit.al [hunk 19]
```diff
@@ -58,6 +65,253 @@ codeunit 148333 "Expense Agent Setup API Test"
...
+
+        // [WHEN] The E2E test handler initializes the company
+        ExpenseTestHandlerAPI.Initialize();
+
+        // [THEN] The stale Expense Vendor is removed
+        Assert.IsFalse(
+            ExpenseVendor.Get(ExpenseVendorNo),
+            'Expense Vendor must be deleted during test initialization.');
+    end;
+
```
[the remark this answers is not in the document]
Comment (t-prda) [reply src/Apps/W1/ExpenseAgent/test/src/API/ExpenseAgentSetupAPITest.Codeunit.al]: System.Collections.Hashtable.b

**Comment by alexei-dobriansky** [comment]
## Good Sense Reviewer - Round 3

# **Recommendation: Accept**

<details>
<summary><strong>What this PR does</strong></summary>

The new commits finish the service-enabled Expense Agent E2E setup API by making Agent state handling consistent and by tightening the setup tests. The setup action now treats a disabled underlying Agent as incomplete, validates the Agent setup state before saving, and keeps the test confirmation expectations isolated per test.

This addresses the remaining correctness issue from the previous round. The enabled-state path now uses the same validated state transition that `Agent Setup` expects, and the completeness check no longer returns success while the Agent is disabled.
</details>

<details>
<summary><strong>Status of previous suggestions</strong></summary>

| ID | Title | Status | Author response |
|----|-------|--------|-----------------|
| S1 | Existing disabled Agent stays disabled | Addressed | The reply says the disabled-Agent path was changed and tested. The current diff confirms this: state changes now use `Validate(State, Enabled)`, the completion check rejects a disabled Agent, and a test covers re-enabling an existing disabled Agent. |
</details>

<details>
<summary><strong>New observations (commits since round 2)</strong></summary>

*None - the new commits address the previous Agent state issue and adjust tests without introducing another issue.*
</details>

<details>
<summary><strong>Risk assessment and necessity</strong></summary>

**Risk:** The changed surface is E2E test setup infrastructure, not posting logic, so direct financial risk is low. The main regression surface is test reliability for companies that already have Expense Agent setup or an existing Agent record; the new state validation and setup-complete check reduce that risk.

**Necessity:** The setup API is needed for repeatable service-driven Expense Agent E2E tests. The latest changes are also needed because the API must keep the setup record and the underlying Agent state in sync before later E2E flows run.
</details>

---
> [AI-PR-REVIEW] version=1 promptVersion=4 system=github pr=10846 round=3 by=alexei-dobriansky at=2026-09-01T13:20:19Z lastSha=c79f1f56a81cf19b483ddcfdd3d0185e0f561670 reviewKey=3888ee4d26975ecdaf1027d882bb50ece5968194105668f54ad75b89375cf457 suggestions=S1@f54b7baa:addressed parentRound=2

File: src/Apps/W1/ExpenseAgent/test/src/Helper/ExpenseTestHandlerAPI.Codeunit.al [hunk 20]
```diff
@@ -26,6 +81,63 @@ codeunit 148307 "Expense Test Handler API"
         exit('Initialize completed');
     end;
 
+    /// <summary>
+    /// Explicitly provisions and enables Expense Agent setup for E2E tests.
+    /// </summary>
+    /// <returns>A completion message after the Expense Agent setup is configured.</returns>
+    [ServiceEnabled]
```
[the remark this answers is not in the document]
Comment (t-prda) [reply src/Apps/W1/ExpenseAgent/test/src/Helper/ExpenseTestHandlerAPI.Codeunit.al:34]: This entry point is contained in the Expense Agent Tests app, not the product app. Publishing that test app to a disposable/local E2E environment is the intentional trust boundary; production/SaaS deployments do not contain this ServiceEnabled codeunit. Adding product entitlement logic to a test-only handler would not strengthen the production boundary.

File: src/Apps/W1/ExpenseAgent/test/src/API/ExpenseAgentSetupAPITest.Codeunit.al [hunk 21]
```diff
@@ -58,9 +65,257 @@ codeunit 148333 "Expense Agent Setup API Test"
...
+            AgentUserSecurityId,
+            "Agent Metadata Provider"::"Expense Agent",
+            '',
+            '',
+            '');
+        TempAgentSetupBuffer.Validate(State, TempAgentSetupBuffer.State::Disabled);
+        AgentSetup.SaveChanges(TempAgentSetupBuffer);
+
+        // [WHEN] The E2E test handler configures the company again
+        ExpenseTestHandlerAPI.Configure();
```
[the remark this answers is not in the document]
Comment (t-prda) [reply src/Apps/W1/ExpenseAgent/test/src/API/ExpenseAgentSetupAPITest.Codeunit.al]: Fixed in 1857b0f0c. The second Configure call now enqueues its own expected confirmation and immediately asserts that the variable-storage queue is empty.

**Comment by alexei-dobriansky** [comment]
## Good Sense Reviewer - Round 4

# **Recommendation: Accept**

<details>
<summary><strong>What this PR does</strong></summary>

The new commits tighten the repeated setup test by queuing the expected confirmation before the second setup call and asserting that the confirmation queue is empty afterward. They also add the new setup API tests to the disabled-test list so the generic BCApps matrix does not run agent-dependent scenarios.

The confirmation change is correct for the disabled-Agent path: the second setup call must run the setup flow again, so the expected prompt must be queued again. The disabled-test entries are scoped to this setup test codeunit's new methods and preserve the tests for explicit local or dedicated pipeline validation without changing runtime app behavior.
</details>

<details>
<summary><strong>Status of previous suggestions</strong></summary>

| ID | Title | Status | Author response |
|----|-------|--------|-----------------|
| S1 | Existing disabled Agent stays disabled | Addressed | No new response. The current diff still confirms the fix: state changes use `Validate(State, Enabled)`, the complete check rejects a disabled underlying Agent, and the disabled-Agent path has a regression test. |
</details>

<details>
<summary><strong>New observations (commits since round 3)</strong></summary>

*None - the new commits queue the second confirmation and disable the agent-dependent setup tests in the public matrix without introducing another issue.*
</details>

<details>
<summary><strong>Risk assessment and necessity</strong></summary>

**Risk:** The changed surface is E2E test setup infrastructure, not posting or production logic, so direct customer and financial risk is low. The main regression surface is test reliability: the new confirmation assertion reduces false positives, while the disabled-test entries mean coverage depends on the dedicated Expense Agent validation path rather than the generic BCApps matrix.

**Necessity:** The setup API remains necessary for repeatable service-driven Expense Agent E2E tests. The latest changes are also necessary because the disabled-Agent test needs its own confirmation expectation, and the agent-dependent setup tests should not run in the generic country test matrix.
</details>

---
> [AI-PR-REVIEW] version=1 promptVersion=4 system=github pr=10846 round=4 by=alexei-dobriansky at=2026-09-01T19:18:41.4951360Z lastSha=20d924c3f5fe0922a4abde0dc0a9e2a9965a8b38 reviewKey=4a78ccf8521c51239b3c56843cd1594d1908db0c2702729c5e2b26f531008472 suggestions=S1@f54b7baa:addressed parentRound=3

File: src/DisabledTests/Expense_Agent_Tests/Expense_Agent_Tests.DisabledTest.json [hunk 22]
```diff
@@ -64,6 +64,31 @@
     "codeunitName": "Expense Agent Setup API Test",
     "method": "EmailAddressFieldIsExposedThroughAPI"
   },
+  {
```
[the remark this answers is not in the document]
Comment (t-prda) [reply src/DisabledTests/Expense_Agent_Tests/Expense_Agent_Tests.DisabledTest.json]: Addressed by 139a45ec0: the five new AL API tests and their disabled-list entries were removed from this PR. Their implementation and tracking are explicitly deferred to AB#636056, while this PR retains only the ServiceEnabled handler needed by the service E2E pipeline.

**Comment by alexei-dobriansky** [comment]
## Good Sense Reviewer - Round 5

# **Recommendation: Accept with Suggestions**

<details>
<summary><strong>What this PR does</strong></summary>

The new commits defer the setup API tests out of this PR. The current net change is the service-enabled Configure action for Expense Agent E2E setup, the enabled-Agent helper logic, and cleanup of stale Expense Vendor records between runs.

The core setup path still looks correct: defaults are created before enabling the agent, existing disabled Agents are re-enabled through Validate(State, Enabled), and the complete check rejects a disabled underlying Agent before short-circuiting. The latest removals do not change that production test-helper logic, but they do reduce the local regression coverage that proved those paths.
</details>

<details>
<summary><strong>Status of previous suggestions</strong></summary>

| ID | Title | Status | Author response |
|----|-------|--------|-----------------|
| S1 | Existing disabled Agent stays disabled | Addressed | No new response. The current helper still validates the Agent setup buffer state before saving, and the complete check still rejects a disabled underlying Agent before returning success. |
</details>

<details>
<summary><strong>New observations (commits since round 4)</strong></summary>

**S2** (🟠 Moderate): *Configure path no longer has local AL coverage*
The setup tests that exercised Configure() were removed from this PR. This API is test setup infrastructure, not posting or data-integrity logic, so this is not a blocker. A small local test for idempotent setup and disabled-Agent recovery would still protect the helper from future regressions.
</details>

<details>
<summary><strong>Risk assessment and necessity</strong></summary>

**Risk:** The changed surface is E2E test setup infrastructure, not posting or production business logic, so direct customer and financial risk is low. The main risk is test reliability: without local AL coverage for the new setup action, a future change can break repeatable setup while only the dedicated service validation catches it.

**Necessity:** The setup API is needed for repeatable service-driven Expense Agent E2E tests, and the current scope is aligned with that goal. The latest commits intentionally narrow this PR to the helper surface and move scenario tests out of the current diff, which is acceptable if that follow-up remains tracked.
</details>

---
> [AI-PR-REVIEW] version=1 promptVersion=4 system=github pr=10846 round=5 by=alexei-dobriansky at=2026-09-02T01:12:51.2013903Z lastSha=139a45ec0f38e65fed4c00150ac3cf62d4381c59 reviewKey=12902ad30bc35995503f9aab6627896dca1f9b6ac7c13233361ab7092658d914 suggestions=S1@f54b7baa:addressed,S2@3d46eb35:new parentRound=4

**Comment by alexei-dobriansky** [comment]
## Good Sense Reviewer - Round 6

# **Recommendation: Accept with Suggestions**

<details>
<summary><strong>What this PR does</strong></summary>

The new commit moves the Expense Agent activation helper out of the service-enabled test handler and into the shared Expense test library. The setup API still creates required defaults, enables or reuses the Agent, and treats the setup as complete only when both the setup record and the underlying Agent state are enabled.

This move keeps `Configure()` focused on orchestration and makes the activation logic reusable from tests. The current code still uses `Validate(State, Enabled)` before saving Agent setup changes, so the previously risky disabled-Agent path remains fixed.
</details>

<details>
<summary><strong>Status of previous suggestions</strong></summary>

| ID | Title | Status | Author response |
|----|-------|--------|-----------------|
| S1 | Existing disabled Agent stays disabled | Addressed | No new response. The current code still validates the Agent setup buffer state before saving, and the complete check still rejects a disabled underlying Agent before returning success. |
| S2 | Configure path no longer has local AL coverage | Not addressed | No new response. The current diff still has no local AL test that exercises the `Configure()` setup path. |
</details>

<details>
<summary><strong>New observations (commits since round 5)</strong></summary>

*None - the new commit only moves helper code into the test library and does not introduce another issue.*
</details>

<details>
<summary><strong>Risk assessment and necessity</strong></summary>

**Risk:** The changed surface is E2E test setup infrastructure, not posting or production business logic, so direct customer and financial risk is low. The remaining regression surface is test reliability for existing setup and disabled Agent states; the open non-blocking risk is that no local AL test protects `Configure()` from future changes.

**Necessity:** The setup API is needed for repeatable service-driven Expense Agent E2E tests. Moving the activation helper to the test library is appropriately scoped because it is test infrastructure and not a production API surface.
</details>

---
> [AI-PR-REVIEW] version=1 promptVersion=4 system=github pr=10846 round=6 by=alexei-dobriansky at=2026-09-02T19:14:47Z lastSha=23c57a1ed855f5cc18eb8aba8f4907fb2ff87f62 reviewKey=ac179973e83c3fe7a57cad9e6def197716b855f4895fb02a313a42967e833d4a suggestions=S1@f54b7baa:addressed,S2@3d46eb35:notaddressed parentRound=5

**Review by Alexander-Ya — APPROVED** [review]

**Review by haoranpb — APPROVED** [review]
